Added global settings for configurable suppressed databus events

billkalter opened the pull request:
## Github Issue #

[26](https://github.com/bazaarvoice/emodb/issues/26)

## What Are We Doing Here?

Three changes:

* The condition for ignorable suppressed databus events is now configurable rather than hard-coded to `not({..,"~tags",contains("re-etl")})`.
* A new DW task was created for updating the condition.
* A re-usable settings-management system was created for storing and propagating the condition through the system.

## How to Test and Verify

1. Create a databus subscription:

   ```
   $ curl -s -XPUT -H "Content-Type: application/x.json-condition" 'http://localhost:8080/bus/1/test-sub?ignoreSuppressedEvents=true' --data-binary 'alwaysTrue()'
   {"success":true}
   ```
2. Assuming a clean system there should be not suppressable events configured:

   ```
   $ curl -s 'http://localhost:8080/bus/1/test-sub'
   {"name":"test-sub","tableFilter":"alwaysTrue()","expiresAt":"2016-09-03T15:47:51.146Z","eventTtl":86400000}
   ```
3. Change the global condition:

   ```
   $ curl -XPOST 'localhost:8081/tasks/databus-suppressed-event-condition?value=\{..,"~tags":containsAny("ignore")\}'
   Prior value:   alwaysFalse()
   Updated value: {..,"~tags":contains("ignore")}
   ```
4. Renew the subscription; it should now contain the suppressed condition:

   ```
   $ curl -s -XPUT -H "Content-Type: application/x.json-condition" 'http://localhost:8080/bus/1/test-sub?ignoreSuppressedEvents=true' --data-binary 'alwaysTrue()'
   {"success":true}
   $curl -s 'http://localhost:8080/bus/1/test-sub'
   {"name":"test-sub","tableFilter":"and(alwaysTrue(),not({..,\"~tags\":contains(\"ignore\")}))","expiresAt":"2016-09-03T15:53:38.180Z","eventTtl":86400000}
    ```

## Risk

Fairly low; the area of the system affected is very localized, and the more complicated code around inserting the suppressed event condition was already in place.

### Level 

Medium

### Required Testing

Manual

### Risk Summary

This change doesn't really add many new risks.  The largest is the uncertainty due to natural delays.  Cross-data-center replication of the setting may take up to a minute, longer if there is a break in the
Cassandra replication.  Additionally, the ignored events are only updated per-subscription at renew time.  So, for example, if a subscription only renews once every four hours then updating the suppressed event condition immediately may take up to four hours to go into effect for that subscription.  Not really a bug, just something to keep in mind when changing the condition; the administrator probably wants to make the update with ample lead time before anyone floods the system with events that should be suppressed.

## Code Review Checklist

- [ ] Tests are included. If not, make sure you leave us a line or two for the reason.

- [ ] Pulled down the PR and performed verification of at least being able to
build and run.

- [ ] Well documented, including updates to any necessary markdown files. When
we inevitably come back to this code it will only take hours to figure out, not
days.

- [ ] Consistent/Clear/Thoughtful? We are better with this code. We also aren't
a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.

- [ ] PR has a valid summary, and a good description.

fahdsiddiqui commented on web/src/main/java/com/bazaarvoice/emodb/web/settings/DatabusSuppressedEventConditionAdminTask.java:
```diff
@@ -0,0 +1,42 @@
+import com.google.inject.Inject;
+
+/**
+ * Task for managing the condition used to suppress databus events when a subscription opts to ignore suppressed events.
```
Should we call it a "suppressed" event condition? Or simply a default "augmented" condition. Reason is it may get confusing dealing with double negatives. Making it more generic may take the confusion away.

fahdsiddiqui commented on web/src/main/java/com/bazaarvoice/emodb/web/settings/DatabusSuppressedEventConditionAdminTask.java:
```diff
@@ -0,0 +1,42 @@
+ *
+ * Update the condition condition:
+ * <code>
+ *     $ curl -XPOST 'localhost:8081/tasks/databus-suppressed-event-condition?value=\{..,"~tags":contains("re-etl")\}'
```
Based on previous comment, it may be confusing to add an "exclusion" condition without adding a "not" to the provided condition.

fahdsiddiqui commented on web/src/main/java/com/bazaarvoice/emodb/web/settings/SettingsManager.java:
```diff
@@ -0,0 +1,278 @@
+        };
+    }
+
+    private final class SettingImpl<T> implements Setting<T> {
```
I would give this class a file of its own for easy readability.

fahdsiddiqui commented on web/src/main/java/com/bazaarvoice/emodb/web/settings/Setting.java:
```diff
@@ -0,0 +1,47 @@
+     * center after the method returns; remote data centers will eventually see the updated value, typically in less
+     * than one minute.
+     */
+    void set(@Nullable T value);
```
Not sure if we should allow the value to be nullable. Thoughts?

fahdsiddiqui commented:
@billkalter - just a few comments. Like the Settings store idea specifically for A/B testing use case.

billkalter commented:
@fahdsiddiqui Addressed feedback.  Specificially:

- Changed settings to enforce a non-null default
- Changed task and setting from negative condition to positive condition
- Renamed setting annotation from `SuppressedEventCondition` to `DefaultJoinFilter`
  - As a side effect if this for consistency I changed the parameter name from "ignoreSuppressedEvents" to "includeDefaultJoinFilter".  In `DatabusResource` I kept support for the old parameter name for backwards compatibility.
